src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,920 @@
+            }
+
+            // TODO remove this method: should be split into 2 or refactored
+            private bool TryConvertQueryExpression(
+                Func<ExpressionSyntax, StatementSyntax> leafExpressionCreationMethod,
+                out StatementSyntax[] statements)
+            {
+                statements = default;
+                if (TryCreateStackFromQueryExpression(out var stack))
+                {
+                    statements = GenerateStatements(leafExpressionCreationMethod, stack);
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool TryProcessQueryBody(QueryBodySyntax queryBody, QueryExpressionProcessingInfo queryExpressionProcessingInfo)
+            {
+                foreach (var queryClause in queryBody.Clauses)
+                {
+                    switch (queryClause.Kind())
+                    {
+                        case SyntaxKind.WhereClause:
+                            queryExpressionProcessingInfo.Add(queryClause);
+                            break;
+                        case SyntaxKind.LetClause:
+                            if (!queryExpressionProcessingInfo.TryAdd(queryClause, ((LetClauseSyntax)queryClause).Identifier))
+                            {
+                                return false;
+                            }
+
+                            break;
+                        case SyntaxKind.FromClause:
+                            var fromClause = (FromClauseSyntax)queryClause;
+                            if (!queryExpressionProcessingInfo.TryAdd(queryClause, fromClause.Identifier))
+                            {
+                                return false;
+                            }
+
+                            break;
+                        case SyntaxKind.JoinClause:
+                            var joinClause = (JoinClauseSyntax)queryClause;
+                            if (joinClause.Into == null) // GroupJoin is not supported
+                            {
+                                if (queryExpressionProcessingInfo.TryAdd(joinClause, joinClause.Identifier))
+                                {
+                                    break;
+                                }
+                            }
+
+                            return false;
+                        // OrderBy is not supported by foreach.
+                        default:
+                            return false;
+                    }
+                }
+
+                // GroupClause is not supported by the conversion
+                if (queryBody.SelectOrGroup.Kind() != SyntaxKind.SelectClause)
+                {
+                    return false;
+                }
+
+                queryExpressionProcessingInfo.Add(queryBody.SelectOrGroup);
+                return queryBody.Continuation == null || TryProcessQueryBody(queryBody.Continuation.Body, queryExpressionProcessingInfo);
```

Comment:
>TryProcessQueryBody [](start = 57, length = 19)

Consider if it would be worth avoiding recursion here. A simple way to do that is to ```queryBody = queryBody.Continuation.Body``` and goto to the ```foreach```. Can also be done with an addition of a ```do while``` loop that walks the list of QueryBodySyntax nodes.